### PR TITLE
fix grpo and add gen logprobs

### PR DESCRIPTION
**Pull request by samsja**

### CODE REVIEW

File: src/zeroband/training/loss.py [hunk 1]
```diff
@@ -25,52 +26,47 @@ def grpo_loss(
...
+    """
+    Computes the GRPO loss.
+    """
+    # Get log probs from current policy
+    log_probs = F.log_softmax(logits, dim=-1)
 
-    # Policy loss
-    policy_loss = -torch.min(ratio * advantages, clipped_ratio * advantages)
+    # Extract per-token log probs for the tokens that were generated
+    per_token_logps = torch.gather(log_probs, dim=2, index=input_ids.unsqueeze(-1)).squeeze(-1)
```
Comment (Jackmin801) [inline src/zeroband/training/loss.py]: Would we save memory if we did `sparse_grad=True`? https://pytorch.org/docs/main/generated/torch.gather.html

File: src/zeroband/training/loss.py [hunk 2]
```diff
@@ -25,52 +26,47 @@ def grpo_loss(
...
+    per_token_loss = torch.min(unclipped_loss, clipped_loss)
 
-    # Final loss (mean across batch)
-    loss = (policy_loss + beta * kl_penalty).mean()
+    # Apply mask and average
+    masked_loss = (per_token_loss * loss_mask).sum(dim=1) / loss_mask.sum(dim=1).clamp(min=1e-5)
 
-    return loss
+    # Return mean loss
+    return -masked_loss.mean()  # Negative because we want to maximize reward
```
Comment (Jackmin801) [inline src/zeroband/training/loss.py]: Maybe move the the negative up to the line where masked_loss is set?

File: src/zeroband/training/loss.py [hunk 3]
```diff
@@ -25,52 +26,47 @@ def grpo_loss(
...
+    """
+    Computes the GRPO loss.
+    """
+    # Get log probs from current policy
+    log_probs = F.log_softmax(logits, dim=-1)
 
-    # Policy loss
-    policy_loss = -torch.min(ratio * advantages, clipped_ratio * advantages)
+    # Extract per-token log probs for the tokens that were generated
+    per_token_logps = torch.gather(log_probs, dim=2, index=input_ids.unsqueeze(-1), sparse_grad=True).squeeze(-1)
```
Comment (Jackmin801) [inline src/zeroband/training/loss.py]: Did you try running the training? Seems like it doesnt work because log_softmax has no sparse bwd impl

image

Comment (Jackmin801) [reply src/zeroband/training/loss.py]: Also, we might want to use the trl one. I think its faster. They are doing selection before the softmax whis is less flops

Comment (samsja) [reply src/zeroband/training/loss.py]: done

**Review by Jackmin801 — APPROVED** [review]
lfgtm
